stl/inc/deque

```diff
@@ -1236,52 +1314,84 @@ private:
-        auto _UFirst      = _Get_unwrapped(_First);
-        const auto _ULast = _Get_unwrapped(_Last);
+        if (_First == _Last) {
+            return begin() + static_cast<difference_type>(_Off);
+        }
 
-        size_type _Oldsize = _Mysize();
+        const size_type _Oldsize = _Mysize();
 
-        if (_UFirst != _ULast) {
-            if (_Off <= _Mysize() / 2) { // closer to front, push to front then rotate
-                _Restore_old_size_guard<_Pop_direction::_Front> _Guard{this, _Oldsize};
-                for (; _UFirst != _ULast; ++_UFirst) {
-                    emplace_front(*_UFirst); // prepend flipped
+        _Orphan_all();
+        if (_Off <= _Oldsize / 2) { // closer to front, push to front then rotate
+            _Restore_old_size_guard<_Pop_direction::_Front> _Guard{this, _Oldsize};
+            if constexpr (_Bidi == _Is_bidi::_Yes) {
+                while (_First != _Last) {
+                    _Emplace_front_internal(*--_Last); // prepend in order
                 }
-
-                _Guard._Container = nullptr;
-
-                size_type _Num = _Mysize() - _Oldsize;
-                _STD reverse(begin(), begin() + static_cast<difference_type>(_Num)); // flip new stuff in place
-                _STD rotate(begin(), begin() + static_cast<difference_type>(_Num),
-                    begin() + static_cast<difference_type>(_Num + _Off));
-            } else { // closer to back
-                _Orphan_all();
-                _Restore_old_size_guard<_Pop_direction::_Back> _Guard{this, _Oldsize};
-                for (; _UFirst != _ULast; ++_UFirst) {
-                    _Emplace_back_internal(*_UFirst);
+            } else {
+                for (; _First != _Last; ++_First) {
+                    _Emplace_front_internal(*_First); // prepend flipped
                 }
+            }
 
-                _Guard._Container = nullptr;
+            _Guard._Container = nullptr;
 
-                _STD rotate(begin() + static_cast<difference_type>(_Off),
-                    begin() + static_cast<difference_type>(_Oldsize), end());
+            const auto _Num     = static_cast<difference_type>(_Mysize() - _Oldsize);
+            const auto _Myfirst = _Unchecked_begin();
+            const auto _Mymid   = _Myfirst + _Num;
+            if constexpr (_Bidi == _Is_bidi::_No) {
+                _STD reverse(_Myfirst, _Mymid); // flip new stuff in place
             }
+            _STD rotate(_Myfirst, _Mymid, _Mymid + static_cast<difference_type>(_Off));
+            return begin() + static_cast<difference_type>(_Off);
         }
 
+        _Restore_old_size_guard<_Pop_direction::_Back> _Guard{this, _Oldsize};
+        for (; _First != _Last; ++_First) {
+            _Emplace_back_internal(*_First);
+        }
+
+        _Guard._Container = nullptr;
+
+        _STD rotate(
+            begin() + static_cast<difference_type>(_Off), begin() + static_cast<difference_type>(_Oldsize), end());

```

Comment:
I observe that these could also use `_Unchecked_begin()`, as the rest of the function is being improved like that.